src/persist/src/storage.rs

```diff
@@ -79,6 +82,115 @@ pub trait Blob {
     fn close(&mut self) -> Result<bool, Error>;
 }
 
+/// The partially structured information stored in an exclusive-writer lock.
+///
+/// To allow for restart without operator intervention in situations where we've
+/// crashed and are unable to cleanly unlock (s3 has no flock equivalent, for
+/// example), this supports reentrance. We define the "same" process as any that
+/// have the same reentrance id, which is an opaque user-provided string.
+///
+/// This, in essence, delegates the problem of ensuring writer-exclusivity to
+/// the persist user, which may have access to better (possibly distributed)
+/// locking primitives than we do.
+///
+/// Concretely, MZ Cloud will initially depend on the exclusivity of attaching
+/// an EBS volume. We'll store the reentrant_id somewhere on the EBS volume that
+/// holds the catalog. Any process that starts and has access to this volume is
+/// guaranteed that the previous machine is no longer available and thus the
+/// previous mz process is no longer running. Similarly, a second process cannot
+/// accidentally start up pointed at the same storage locations because it will
+/// not have the reentrant_id available.
+///
+/// Violating writer-exclusivity will cause undefined behavior including data
+/// loss. It's always correct, and MUCH safer, to provide a random reentrant_id
+/// here (e.g. a UUID). It will simply require operator intervention to verify
+/// that the previous process is no longer running in the event of a crash and
+/// to confirm this by manually removing the previous lock.
+#[derive(Debug, Clone)]
+pub struct LockInfo {
+    reentrance_id: String,
+    details: String,
+}
+
+impl LockInfo {
+    /// Returns a new LockInfo from its component parts.
+    pub fn new(reentrance_id: String, details: String) -> Self {
+        LockInfo {
+            reentrance_id,
+            details,
+        }
+    }
+
+    /// Constructs a new, empty LockInfo with a unique reentrance id.
+    ///
+    /// Helper for tests that don't care about locking reentrance (which is most
+    /// of them).
+    pub fn new_no_reentrance(details: String) -> Self {
+        use std::time::{SystemTime, UNIX_EPOCH};
+        let reentrance_id = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_nanos()
+            .to_string();
+        Self::new(reentrance_id, details)
+    }
+
+    /// Returns Ok(Ok()) if this lock information represents a process that may
+    /// proceed in the presence of some existing lock.
+    ///
+    /// The outer result represents success or failure in locking the previous
+    /// lock's information. The inner result represents the ability to proceed
+    /// or, in the event of an existing lock we are not reentrant for, the
+    /// contents of the previous lock (for use in an error message).
+    pub fn is_reentrant_for<R: Read>(&self, mut existing: R) -> Result<Result<(), String>, Error> {
```

Comment:
I think it might be simpler / harder to misuse this API if it just returned a `Result<(), Error>>` instead of nested results (and maybe we make the Error variants more descriptive as a followup?